### PR TITLE
mds: add perf counters for file system operations

### PR DESCRIPTION
**Pull request by michaelsevilla**
Shows the types of operations that clients are doing by adding counters to the
spot where operations are demuxed. Does not support per-client activity. This
is the first step towards comprehensive performance tracking in the file
system.

Signed-off-by: Michael Sevilla <mikesevilla3@gmail.com>

### CODE REVIEW

**Comment by michaelsevilla** [comment]
Addresses: http://tracker.ceph.com/issues/19362

File: src/mds/Server.cc [hunk 1]
```diff
@@ -1533,100 +1587,127 @@ void Server::dispatch_client_request(MDRequestRef& mdr)
...
     handle_client_mknod(mdr);
     break;
   case CEPH_MDS_OP_LINK:
+    logger->inc(l_mdss_req_link);
     handle_client_link(mdr);
     break;
   case CEPH_MDS_OP_UNLINK:
+    logger->inc(l_mdss_req_unlink);
   case CEPH_MDS_OP_RMDIR:
+    logger->inc(l_mdss_req_rmdir);
```
Comment (batrick) [inline src/mds/Server.cc]: `l_mdss_req_rmdir` gets incremented for unlink and rmdir (see fallthrough for unlink).

File: src/mds/Server.cc [hunk 2]
```diff
@@ -1533,100 +1587,127 @@ void Server::dispatch_client_request(MDRequestRef& mdr)
   switch (req->get_op()) {
   case CEPH_MDS_OP_LOOKUPHASH:
   case CEPH_MDS_OP_LOOKUPINO:
+    logger->inc(l_mdss_req_lookupino);
```
Comment (batrick) [inline src/mds/Server.cc]: Let's break these into two counters?

**Comment by michaelsevilla** [comment]
I can't think of any other perf counters I'd want -- I had this code lying around from a research project on load balancing. I have LTTnG code that gives you the latency of each file system request (from the MDS's view), similar to the stuff @noahdesu did [in his blog post on OSD request tracing](http://noahdesu.github.io/2014/06/01/tracing-ceph-with-lttng-ust.html). Would that be useful?

File: src/mds/Server.cc [hunk 3]
```diff
@@ -1586,6 +1588,9 @@ void Server::dispatch_client_request(MDRequestRef& mdr)
 
   switch (req->get_op()) {
   case CEPH_MDS_OP_LOOKUPHASH:
+    logger->inc(l_mdss_req_lookuphash);
+    handle_client_lookup_ino(mdr, false, false);
```
Comment (michaelsevilla) [inline src/mds/Server.cc]: Not sure if you wanted this repeated like this.

Comment (batrick) [inline src/mds/Server.cc]: I don't have a problem with it.

**Comment by batrick** [comment]
@michaelsevilla that sounds very useful! Please make a tracker ticket to collect feedback for it.

**Comment by batrick** [comment]
`Reviewed-by: Patrick Donnelly <pdonnell@redhat.com>`

**Comment by batrick** [comment]
cephfs integration test run: http://pulpito.ceph.com/pdonnell-2017-05-09_17:36:48-fs-wip-pdonnell-integration-distro-basic-smithi/

**Comment by jcsp** [comment]
The downside to this implementation is that it will count operations multiple times when they pass through dispatch_client_request more than once.  That happens when the operation waits for locks or waits for metadata to load.

Perhaps incrementing the counters in respond_to_request instead?  It would require writing out the big case statement by op type again, but the resulting counters would be more meaningful.

**Comment by michaelsevilla** [comment]
Ah - thanks. I was seeing double counting for my experiments and I couldn't figure out why. Hopefully incrementing in `respond_to_request` fixes that problem.

File: src/mds/Server.cc [hunk 4]
```diff
@@ -1046,6 +1102,93 @@ void Server::respond_to_request(MDRequestRef& mdr, int r)
     mdr->internal_op_finish->complete(r);
     mdcache->request_finish(mdr);
   }
+
+  // add here to avoid counting ops multiple times (e.g., locks, loading)
+  switch(mdr->client_request->get_op()) {
```
Comment (jcsp) [inline src/mds/Server.cc]: Oops, missed this in review -- mdr->client_request may be null (see conditional above), so this breaks when internal ops are executed.

Comment (michaelsevilla) [inline src/mds/Server.cc]: That massive switch statement has internal ops (like snapshot and file layout stuff), right? So we want a conditional like:

```C++
if (mdr->client_request || mdr->internal_op > -1) {
```

I suppose I could split the perf counters up -- but I'm not sure which ones are internal ops.

Comment (jcsp) [inline src/mds/Server.cc]: I think all the ops you've got stats for are client ones (that includes snapshots/layout stuff).  The internal ones are listed in include/ceph_fs.h line 363 onwards.

Comment (jcsp) [inline src/mds/Server.cc]: ...and I think it's fine to not have the counters for internal ops here

**Comment by michaelsevilla** [comment]
Whoops -- bad rebase. Let me fix it.

**Pull request merged**